src/__experimental__/components/date/date.component.js

```diff
@@ -162,34 +180,53 @@ class BaseDateInput extends React.Component {
     if (isValidDate) {
       isoDateString = DateHelper.formatValue(dateWithSlashes);
       this.updateSelectedDate(isoDateString);
-      this.emitOnChangeCallback(isoDateString);
+      this.emitOnChangeCallback(ev, isoDateString);
     }
 
-    this.setState({ visibleValue: ev.target.value });
+    this.setState({ visibleValue: value });
   };
 
   updateSelectedDate = (newValue) => {
-    let newDate = DateHelper.stringToDate(newValue);
+    const newDate = this.getDateObject(newValue);
+
+    this.setState({ selectedDate: newDate });
+  };
+
+  getDateObject = (newValue) => {
+    let newDate = DateHelper.stringToDate(DateHelper.formatValue(newValue));
     const isNewDateInvalid = !newDate.getDate();
 
     if (isNewDateInvalid) {
       newDate = DateHelper.stringToDate(DateHelper.todayFormatted());
     }
 
-    this.setState({ selectedDate: newDate });
+    return newDate;
   };
 
-  emitOnChangeCallback = (stringDate) => {
-    const changePayload = {
-      target: {
-        name: this.props.name,
-        value: stringDate
-      }
-    };
-
-    if (this.props.onChange) this.props.onChange(changePayload);
+  emitOnChangeCallback = (ev, isoFormattedValue) => {
+    if (this.props.onChange) {
+      const event = this.buildCustomEvent(ev, isoFormattedValue);
+      this.props.onChange(event);
+    }
   };
 
+  buildCustomEvent = (ev, isoFormattedValue) => {
+    const { id, name, value } = ev.target;
+    const { lastValidEventValues } = this.state;
+    const validRawValue = DateHelper.isValidDate(isoFormattedValue);
+
+    ev.target = {
+      ...(name && { name }),
+      ...(id && { id }),
+      value:
```

Comment:
this feels bit overwhelming, perhaps we could refactor that into separate helper function?

Reply:
Maybe this would be better?
```js
value: {
  ...lastValidEventValues,
  ...(validRawValue && { formattedValue: formatDateToCurrentLocale(value) }),
  ...(validRawValue && { rawValue: isoFormattedValue })
}
```
Would that work?

Reply:
Yep good point, I had this as WIP whilst I tried it out, I'll refactor it thanks